lib/cfndsl/cloud_formation_template.rb

```diff
@@ -0,0 +1,179 @@
+          if ptype.instance_of?(String)
+            create_klass = type_module.const_get(ptype)
+
+            klass.class_eval do
+              CfnDsl.method_names(pname) do |method|
+                define_method(method) do |*values, &block|
+                  values.push create_klass.new if values.empty?
+
+                  @Properties ||= {}
+                  @Properties[pname] = CfnDsl::PropertyDefinition.new(*values)
+                  @Properties[pname].value.instance_eval(&block) if block
+                  @Properties[pname].value
+                end
+              end
+            end
+          else
+            # Array version
+            sing_name = CfnDsl::Plurals.singularize(pname)
+            create_klass = type_module.const_get(ptype[0])
+            klass.class_eval do
+              CfnDsl.method_names(pname) do |method|
+                define_method(method) do |*values, &block|
+                  values.push [] if values.empty?
+                  @Properties ||= {}
+                  @Properties[pname] ||= PropertyDefinition.new(*values)
+                  @Properties[pname].value.instance_eval(&block) if block
+                  @Properties[pname].value
+                end
+              end
+
+              CfnDsl.method_names(sing_name) do |method|
+                define_method(method) do |value = nil, &block|
+                  @Properties ||= {}
+                  @Properties[pname] ||= PropertyDefinition.new([])
+                  value = create_klass.new unless value
+                  @Properties[pname].value.push value
+                  value.instance_eval(&block) if block
+                  value
+                end
+              end
+            end
+          end
+        end
+        parts = name.split('::')
+        until parts.empty?
+          abreve_name = parts.join('_')
+          if names.key?(abreve_name)
+            # this only happens if there is an ambiguity
+            names[abreve_name] = nil
+          else
+            names[abreve_name] = type_module.const_get(klassname)
+            nametypes[abreve_name] = name
+          end
+          parts.shift
+        end
+      end
+
+      # Define property setter methods for each of the unambiguous type names
+      names.each_pair do |typename, type|
+        next unless type
+
+        class_eval do
+          CfnDsl.method_names(typename) do |method|
+            define_method(method) do |name, *values, &block|
+              name = name.to_s
+              @Resources ||= {}
+              resource = @Resources[name] ||= type.new(*values)
+              resource.instance_eval(&block) if block
+              resource.instance_variable_set('@Type', nametypes[typename])
+              resource
+            end
+          end
+        end
+      end
+    end
+  end
+
+  class CloudFormationTemplate < OrchestrationTemplate
+    def self.template_types
+      CfnDsl::AWS::Types::Types_Internal
```

Comment:
These `Internal` prefixes aren't going to stay, I need to work out a better name for these otherwise it's `.. AWS::Types::Types` which doesn't really make much sense.